### ISSUE TITLE
ensure `pnpm dlx foo@latest` uses the latest version available

### ISSUE THREAD

**Issue by dominikg** [pnpm/pnpm#8659]
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

As a developer I expect running the latest version of foo when executing `pnpm dlx foo@latest` after hearing of a new release. 

but i just ran into an issue that the dlx cache includes caching latest, leading to execution of a stale older version.

### Describe the solution you'd like

at least when using the `latest` tag, (but possibly any tag, because next, beta, alpha are also frequently used). Before executing a locally cached version pnpm dlx should revalidate that this is still the current version for that tag and fetch the new one if it is not.

### Describe the drawbacks of your solution

slight overhead for checking the tag version with an http request

### Describe alternatives you've considered

setting dlx cache max age to 0, bypassing the cache completely and ensuring even more http requests happen

**Comment by zkochan** [comment]
> Before executing a locally cached version pnpm dlx should revalidate that this is still the current version for that tag and fetch the new one if it is not.

The problem with this approach is that we actually allow running dlx with a list of package names. E.g.:

```
pnpm --package=yo --package=generator-webapp dlx yo webapp --skip-install
```

And the cache is a combination of these package names. So, should we resolve all tags in all listed packages?

**Comment by dominikg** [comment]
yes. You can keep the cache key stable and just download those that have changed + update the cache. as a user i feel thats a lot less surprising than getting an older version while using `@latest`

**Comment by dominikg** [comment]
come to think of it, you could even change the caching strategy to version based first, with a possible throttle time to allow users to skip the version lookup (should be minutes rather than a full day by default)

**Comment by LasyIsLazy** [comment]
> come to think of it, you could even change the caching strategy to version based first, with a possible throttle time to allow users to skip the version lookup (should be minutes rather than a full day by default)

I agree with that.

Now, pnpm dlx use the hash of package names as cache key,

https://github.com/pnpm/pnpm/blob/e476b07eb4ac32efd135bb4b809fff67a81e6662/exec/plugin-commands-script-runners/src/dlx.ts#L200-L205

which means, `pnpm dlx foo@latest` will use the `latest` tag as a part of cache key rather than the real version. So, if `latest` tag update, the cache will match a wrong version.

**Comment by dominikg** [comment]
it appears that pnpm@10 implemented this. https://github.com/pnpm/pnpm/pull/8811

thanks!
